runtime/compiler/runtime/CRRuntime.cpp

```diff
@@ -0,0 +1,627 @@
+                                     "At Checkpoint Restore:: Swap Memory is %s",
+                                     getCompInfo()->isSwapMemoryDisabled()? "disabled":"enabled");
+      }
+
+   if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestore))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_CHECKPOINT_RESTORE, "Ready for restore");
+   }
+
+void
+TR::CRRuntime::process()
+   {
+   acquireCRRuntimeMonitor();
+   do
+      {
+      while (getCRRuntimeThreadLifetimeState() == TR_CRRuntimeThreadLifetimeStates::CR_THR_INITIALIZED)
+         {
+         waitOnCRRuntimeMonitor();
+         }
+
+      if (getCRRuntimeThreadLifetimeState() == TR_CRRuntimeThreadLifetimeStates::CR_THR_STOPPING)
+         {
+         releaseCRRuntimeMonitor();
+         break;
+         }
+      }
+   while (true);
+   }
+
+static int32_t J9THREAD_PROC crRuntimeThreadProc(void * entryarg)
+   {
+   J9JITConfig *jitConfig = (J9JITConfig *) entryarg;
+   J9JavaVM *vm = jitConfig->javaVM;
+   TR::CompilationInfo *compInfo = TR::CompilationInfo::get(jitConfig);
+   TR::CRRuntime *crRuntime = compInfo->getCRRuntime();
+   J9VMThread *crRuntimeThread = NULL;
+
+   PORT_ACCESS_FROM_JITCONFIG(jitConfig);
+
+   int rc = vm->internalVMFunctions->internalAttachCurrentThread(vm, &crRuntimeThread, NULL,
+                                  J9_PRIVATE_FLAGS_DAEMON_THREAD | J9_PRIVATE_FLAGS_NO_OBJECT |
+                                  J9_PRIVATE_FLAGS_SYSTEM_THREAD | J9_PRIVATE_FLAGS_ATTACHED_THREAD,
+                                  crRuntime->getCRRuntimeOSThread());
+
+   crRuntime->getCRRuntimeMonitor()->enter();
+   if (rc == JNI_OK)
+      {
+      crRuntime->setCRRuntimeThread(crRuntimeThread);
+      j9thread_set_name(j9thread_self(), "CR Runtime");
+      crRuntime->setCRRuntimeThreadLifetimeState(TR::CRRuntime::CR_THR_INITIALIZED);
+      }
+   else
+      {
+      crRuntime->setCRRuntimeThreadLifetimeState(TR::CRRuntime::CR_THR_FAILED_TO_ATTACH);
+      }
+   crRuntime->getCRRuntimeMonitor()->notifyAll();
+   crRuntime->getCRRuntimeMonitor()->exit();
+
+   // attaching the CR Runtime Thread failed
+   if (rc != JNI_OK)
+      return JNI_ERR;
+
+   crRuntime->process();
+
+   vm->internalVMFunctions->DetachCurrentThread((JavaVM *) vm);
+   crRuntime->setCRRuntimeThread(NULL);
+   crRuntime->getCRRuntimeMonitor()->enter();
+   crRuntime->setCRRuntimeThreadLifetimeState(TR::CRRuntime::CR_THR_DESTROYED);
+   crRuntime->getCRRuntimeMonitor()->notifyAll();
+   j9thread_exit((J9ThreadMonitor*)crRuntime->getCRRuntimeMonitor()->getVMMonitor());
+
+   return 0;
+   }
+
+void
+TR::CRRuntime::startCRRuntimeThread(J9JavaVM *javaVM)
+   {
+   PORT_ACCESS_FROM_JAVAVM(javaVM);
+
+   //256KB stack size
+   const UDATA defaultOSStackSize = javaVM->defaultOSStackSize;
+   UDATA priority = J9THREAD_PRIORITY_NORMAL;
+
+   // create the CR Runtime Thread
+   if(javaVM->internalVMFunctions->createThreadWithCategory(&_crRuntimeOSThread,
```

Comment:
Space needed after `if`

Reply:
Immediately after the thread gets created, it is possible for the crRuntime thread to execute all its code and be stuck in a busy loop in process() with CRRuntimeMonitor in hand and FAILED_TO_ATTACH state.